rig/machine_control/machine_controller.py

```diff
@@ -0,0 +1,770 @@
+            fills = regions.compress_flood_fill_regions(targets)
+
+            # Load the APLX data
+            with open(aplx, "rb+") as f:
+                aplx_data = f.read()
+            n_blocks = ((len(aplx_data) + consts.SCP_DATA_LENGTH - 1) //
+                        consts.SCP_DATA_LENGTH)
+
+            # Perform each flood-fill.
+            for (region, cores) in fills:
+                # Get an index for the nearest neighbour operation
+                pid = self._get_next_nn_id()
+
+                # Send the flood-fill start packet
+                self._send_ffs(pid, region, n_blocks, fr)
+
+                # Send the data
+                self._send_ffd(pid, aplx_data, consts.SARK_DATA_BASE)
+
+                # Send the flood-fill END packet
+                self._send_ffe(pid, app_id, flags, cores, fr)
+
+    @ContextMixin.use_contextual_arguments
+    def send_signal(self, signal, app_id=Required):
+        """Transmit a signal to applications.
+
+        Arguments
+        ---------
+        signal : :py:class:`~rig.machine_control.consts.AppSignal`
+            Signal to transmit.
+        """
+        if signal not in consts.AppSignal:
+            raise ValueError(
+                "send_signal: Cannot transmit signal of type {}".format(signal)
+            )
+
+        # Construct the packet for transmission
+        arg1 = consts.signal_types[signal]
+        arg2 = (signal << 16) | 0xff00 | app_id
+        arg3 = 0x0000ffff  # Meaning "transmit to all"
+        self._send_scp(0, 0, 0, SCPCommands.signal, arg1, arg2, arg3)
+
+    @ContextMixin.use_contextual_arguments
+    def load_routing_table_entries(self, entries, x=Required, y=Required,
+                                   app_id=Required):
+        """Load routing table entries into the router of a SpiNNaker chip.
+
+        Parameters
+        ----------
+        entries : [RoutingTableEntry(...), ...]
+            List of :py:class:`rig.routing_table.RoutingTableEntry`\ s.
+        """
+        count = len(entries)
+
+        # Try to allocate room for the entries
+        rv = self._send_scp(
+            x, y, 0, SCPCommands.alloc_free,
+            (app_id << 8) | consts.AllocOperations.alloc_rtr, count
+        )
+        rtr_base = rv.arg1  # Index of the first allocated entry, 0 if failed
+
+        if rtr_base == 0:
+            raise SpiNNakerRouterError(count, x, y)
+
+        # Determine where to write into memory
+        buf = self.read_struct_field(b"sv", b"sdram_sys", x, y)
+
+        # Build the data to write in, then perform the write
+        data = b""
+        for i, entry in enumerate(entries):
+            # Build the route as a 32-bit value
+            route = 0x00000000
+            for r in entry.route:
+                route |= r
+
+            data += struct.pack("<2H 3I", i, 0, r, entry.key, entry.mask)
+        self.write(buf, data, x, y)
+
+        # Perform the load of the data into the router
+        self._send_scp(
```

Comment:
**Bug** (?): If more than 16 routing entries are to be loaded the SCP packet payload will grow beyond 256 bytes and thus won't be handled correctly by SC&MP.